File: app/src/main/java/org/oar/minesweeper/Hud.java
package org.oar.minesweeper;

import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.View;

import org.oar.minesweeper.control.MainLogic;
import org.oar.minesweeper.control.ScreenProperties;
import org.oar.minesweeper.control.Settings;

public class Hud {
	private static final float FONT_SIZE = 9;
	private static final float MARGIN = 10;

	private float fontHeight;
	private float margin;

	private MainLogic logic;
	private View view;

	private Paint bg,textStyle;

	private Timer timer;
	private int startingTime=0;
	protected String timeStr = "";
	protected boolean showTime = false;

	public Hud(MainLogic logic, View view) {
		this.logic = logic;
		this.view = view;
		showTime = Settings.INSTANCE.getShowTime();

		bg = new Paint();

		textStyle = new Paint(Paint.FILTER_BITMAP_FLAG);
		textStyle.setColor(Color.WHITE);
		textStyle.setTextSize(ScreenProperties.fontSizeAdapted(FONT_SIZE));

		Rect r = new Rect();
		textStyle.getTextBounds("W", 0, 1, r);

		fontHeight = r.height();
		margin = ScreenProperties.dpiValue(MARGIN);
	}

	public void draw(Canvas canvas) {
		boolean showTime = this.showTime;
		switch(logic.getStatus()){
			case PLAYING: bg.setColor(0xA0000000); break;
			case LOSE:
				bg.setColor(0xA0FF0000);
				showTime = true;
				break;
			case WIN:
				bg.setColor(0xA000FF00);
				showTime = true;
				break;
		}

		canvas.drawRect(0, ScreenProperties.INSTANCE.getHEIGHT()-margin*2-fontHeight,
				ScreenProperties.INSTANCE.getWIDTH(), ScreenProperties.INSTANCE.getHEIGHT(), bg);

		textStyle.setTextAlign(Paint.Align.LEFT);
		canvas.drawText("Open: "+(logic.getRevisedTiles())+"/"+logic.getTotalTiles(),
				margin, ScreenProperties.INSTANCE.getHEIGHT()-margin, textStyle);

		textStyle.setTextAlign(Paint.Align.RIGHT);
		String rightText = "";
		if (showTime) {
			rightText = timeStr + "   ";
		}
		rightText += logic.getFlaggedBombs() + "/" + logic.getGrid().getBombs();

		canvas.drawText(rightText, ScreenProperties.INSTANCE.getWIDTH()-margin,
				ScreenProperties.INSTANCE.getHEIGHT()-margin, textStyle);
	}

	public void startTimer(){
		if (timer!=null){
			timer.timerClose();
		}
		if (startingTime==0) {
			timer = new Timer();
		} else {
			timer = new Timer(startingTime);
		}
		timer.start();
	}

	public void resumeTimer(){
		showTime = Settings.INSTANCE.getShowTime();
		if (timer==null) return;
		timer.timeResume();
	}
	public void pauseTimer(){
		if (timer==null) return;
		timer.timePause();
	}
	public void restartTimer() {
		if (timer==null){
			startingTime=0;
			startTimer();
		} else {
			timer.restart();
		}
	}
	public void stopTimer() {
		if (timer==null) return;
		timer.timerClose();
		timer = null;
	}
	public int getTime(){
		return timer.getTime();
	}
	public void setTime(int seconds){
		if (timer==null)
			startingTime = seconds;
		else
			timer.setTime(seconds);
	}
	public void timerNotify(int time){
		int s = time / 10;
		int m = s / 60;
		int h = m / 60;
		s %= 60;
		m %= 60;
		timeStr = (h == 0 ? "" : (h + ":")) + m + ":" + (s < 10 ? "0" + s : s);
		if (showTime) {
			view.postInvalidate(0, (int) (ScreenProperties.INSTANCE.getHEIGHT()-margin-fontHeight),
					ScreenProperties.INSTANCE.getWIDTH(), ScreenProperties.INSTANCE.getHEIGHT());
		}
	}

	private class Timer extends Thread{
		long initTime=0,pausedTime=0;
		int initDSecs=0;
		boolean running = true;
		public Timer(){}
		public Timer(int dseconds){
			this.initDSecs = dseconds;
		}
		public int getTime(){
			return (int)((System.nanoTime()-initTime)/100000000);
		}
		public void setTime(int dseconds){
			initTime = System.nanoTime()-dseconds*100000000L;
			pausedTime=0;
		}
		public void restart(){
			running = true;
			initTime = System.nanoTime();
			initDSecs = 0;
			pausedTime = 0;
		}
		public void timePause(){
			if (pausedTime==0){
				pausedTime=System.nanoTime();
			}
		}
		public void timeResume(){
			if (pausedTime!=0) {
				initTime += (System.nanoTime() - pausedTime);
			}
			pausedTime=0;
		}
		public void timerClose(){
			running = false;
		}
		@Override
		public void run() {
			super.run();
			running = true;
			initTime = System.nanoTime()-(initDSecs*100000000L);
			pausedTime=0;
			while (running){
				if (pausedTime==0){
					Hud.this.timerNotify(getTime());
				}
				try {
					Thread.sleep(1000);
				} catch (InterruptedException e) {
					e.printStackTrace();
				}
			}
		}
	}
}

File: app/src/main/java/org/oar/minesweeper/Hud.kt
package org.oar.minesweeper

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.view.View
import androidx.appcompat.content.res.AppCompatResources
import org.oar.minesweeper.control.MainLogic
import org.oar.minesweeper.control.MainLogic.GameStatus
import org.oar.minesweeper.control.ScreenProperties.BUTTON_PANEL_HEIGHT
import org.oar.minesweeper.control.ScreenProperties.HEIGHT
import org.oar.minesweeper.control.ScreenProperties.WIDTH
import org.oar.minesweeper.control.ScreenProperties.adaptFontSize
import org.oar.minesweeper.control.ScreenProperties.toDpi
import org.oar.minesweeper.control.Settings
import kotlin.math.roundToInt


class Hud(
    private val logic: MainLogic,
    private val view: View,
    private val context: Context
) {

    companion object {
        private const val FONT_SIZE = 9
        private const val MARGIN = 10
    }

    private val fontHeight: Float
    private val margin: Float
    private val textStyle = Paint(Paint.FILTER_BITMAP_FLAG)
    private var timer: Timer? = null
    private var startingTime = 0
    private var timeStr = ""
    private var showTime = Settings.showTime

    var time: Int
        get() = timer?.time ?: -1
        set(seconds) {
            timer
                ?.apply { time = seconds }
                ?: run { startingTime = seconds }
        }

    init {
        textStyle.color = Color.WHITE
        textStyle.textSize = FONT_SIZE.adaptFontSize()

        val r = Rect()
        textStyle.getTextBounds("W", 0, 1, r)
        fontHeight = r.height().toFloat()
        margin = MARGIN.toDpi()
    }

    fun draw(canvas: Canvas) {
        var drawableRes = R.drawable.hud_background
        val showTime = when (logic.status) {
            GameStatus.PLAYING -> {
                this.showTime
            }
            GameStatus.LOSE -> {
                drawableRes = R.drawable.hud_background_lose
                true
            }
            GameStatus.WIN -> {
                drawableRes = R.drawable.hud_background_win
                true
            }
        }

        val height = HEIGHT + BUTTON_PANEL_HEIGHT
        AppCompatResources.getDrawable(context, drawableRes)
            ?.also {
                it.setBounds(
                    0,
                    height - margin.roundToInt() * 2 - fontHeight.roundToInt(),
                    WIDTH,
                    height,
                )
                it.draw(canvas)
            }


        textStyle.textAlign = Paint.Align.LEFT
        canvas.drawText(
            "Open: " + logic.revisedTiles + "/" + logic.totalTiles,
            margin,
            height - margin,
            textStyle
        )

        textStyle.textAlign = Paint.Align.RIGHT
        var rightText = if (showTime) "$timeStr   " else ""
        rightText += logic.flaggedBombs.toString() + "/" + logic.grid.bombs
        canvas.drawText(
            rightText,
            WIDTH - margin,
            height - margin,
            textStyle
        )
    }

    fun startTimer() {
        timer?.timerClose()
        timer = Timer(startingTime)
            .also { it.start() }
    }

    fun resumeTimer() {
        showTime = Settings.showTime
        timer?.timeResume()
    }

    fun pauseTimer() {
        timer?.timePause()
    }

    fun restartTimer() {
        timer
            ?.apply { restart() }
            ?: run {
                startingTime = 0
                startTimer()
            }
    }

    fun stopTimer() {
        timer?.timerClose()
        timer = null
    }

    fun timerNotify(time: Int) {
        var ss = time / 10
        var mm = ss / 60
        val hh = mm / 60
        ss %= 60
        mm %= 60

        timeStr = (if (hh == 0) "$mm:" else "$hh" + if (mm < 10) "0$mm" else mm) +
            if (ss < 10) "0$ss" else ss

        if (showTime) {
            view.postInvalidate(
                0,
                (HEIGHT - margin - fontHeight).toInt(),
                WIDTH,
                HEIGHT
            )
        }
    }

    private inner class Timer(
        var initDSecs: Int
    ) : Thread() {

		var initTime: Long = 0
		var pausedTime: Long = 0
		var running = true

        var time: Int
            get() = ((System.nanoTime() - initTime) / 100000000).toInt()
            set(dSeconds) {
                initTime = System.nanoTime() - dSeconds * 100000000L
                pausedTime = 0
            }

        fun restart() {
            running = true
            initTime = System.nanoTime()
            initDSecs = 0
            pausedTime = 0
        }

        fun timePause() {
            if (pausedTime == 0L) {
                pausedTime = System.nanoTime()
            }
        }

        fun timeResume() {
            if (pausedTime != 0L) {
                initTime += System.nanoTime() - pausedTime
            }
            pausedTime = 0
        }

        fun timerClose() {
            running = false
        }

        override fun run() {
            super.run()
            running = true
            initTime = System.nanoTime() - initDSecs * 100000000L
            pausedTime = 0
            while (running) {
                if (pausedTime == 0L) {
                    timerNotify(this.time)
                }
                try {
                    sleep(1000)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }
            }
        }
    }
}